pyccel/ast/variable.py

```diff
@@ -765,15 +765,15 @@ def __init__(
                     stop  = a.stop if a.stop is not None else s
                     step  = a.step
                     if isinstance(start, PyccelUnarySub):
-                        start = PyccelAdd(s, start)
+                        start = PyccelAdd(s, start) # cannot simplify
                     if isinstance(stop, PyccelUnarySub):
-                        stop = PyccelAdd(s, stop)
+                        stop = PyccelAdd(s, stop) # cannot simplify
 
-                    _shape = stop if start is None else PyccelMinus(stop, start, simplify=True)
+                    _shape = stop if start is None else PyccelMinus(stop, start) # cannot simplify
                     if step is not None:
                         if isinstance(step, PyccelUnarySub):
                             start = s if a.start is None else start
-                            _shape = start if a.stop is None else PyccelMinus(start, stop, simplify=True)
+                            _shape = start if a.stop is None else PyccelMinus(start, stop) # cannot simplify
```

Comment:
Can you explain why these can't be simplified? I can't see anything obvious

Reply:
Did you get rid of this due to the errors [here](https://github.com/pyccel/pyccel/runs/2326040155) ? I've just had a look at what is generated. It's not a problem with your code. It's a problem with how we're printing the shape and how we're handling the shape when it's reassigned. It would be better to mark the test as xfail I think. I would open an issue for it, but the error is hidden without this improvement

Reply:
To be clear the test is translating:
```
def reassigned_pointer():
    from numpy import array, shape
    a = array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
    c = a[1:]
    c[1] = 0
    c = a
    return c[0], c[1], c[2], shape(c)[0]
```
In this code the shape of `c` changes. First it is `shape(a)-1=10-1=9`, then it is `shape(a)=10`. Before the improvement above we had `shape(c) = (10-1)` so we decided it was too complicated to print the shape, and we printed `c.shape` instead (which means the code works even if we are using the wrong shape internally). After your fix, the shape is `9` which is simple enough that we print it as is and so the problem becomes apparent

Reply:
It was this that generated the last test fails in my last commit.
Which was the case in `test_arrays_view` function in `tests/pyccel/test_pyccel.py`
while trying to view the array in this function in fortran

```python
def array_view_3():
    a = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
    b = a[1:-1]
    c = b[1:-1]
    c = a
    for i in range(np.shape(c)[0]):
        print(c[i])
```

without simplification the loop looks like this 
```fortran
do i = 0_C_INT64_T, size(c, kind=C_INT64_T) - 1_C_INT64_T, 1_C_INT64_T
  print *, c(i)
end do
```
but with simplification it looks like this
```fortran
do i = 0_C_INT64_T, 5_C_INT64_T, 1_C_INT64_T
  print *, c(i)
end do
```
Which is weird. It looks like it always take the size of c when it was first declared.

Reply:
I found another way to reproduce the bug. See #847

Reply:
> Which is weird. It looks like it always take the size of c when it was first declared.

That's exactly what's happening. Normally you can't change the shape of a `Variable` so there isn't a problem but when you reallocate a pointer you can change the shape

Reply:
I think the best thing to do would be to keep this code and add an `xfail` mark to the failing test referencing issue #847. That way we already have a way to see if #847 is fixed

Reply:
ok

Reply:
This problem should be fixed with PR #851 . You can create a temporary branch combining this branch and branch devel-issue847 to check

Reply:
done in [simplified_nodes_devel_issue847](https://github.com/pyccel/pyccel/tree/simplified_nodes_devel_issue847)

Reply:
Great if those tests pass and you're happy with the solution then you can approve PR #851